src/Microsoft.Data/DataFrame.cs

```diff
@@ -402,6 +402,129 @@ public DataFrame FillNulls(IList<object> values, bool inPlace = false)
             return ret;
         }
 
+        private void ResizeByOneAndAppend(DataFrameColumn column, object value)
+        {
+            long length = column.Length;
+            column.Resize(length + 1);
+            column[length] = value;
+        }
+
+        /// <summary> 
+        /// Appends a row inplace to the DataFrame 
+        /// </summary> 
+        /// <remarks>If a column's value doesn't match its column's data type, a conversion will be attempted</remarks> 
+        /// <remarks>If <paramref name="row"/> is null, a null value is appended to each column</remarks>
+        /// <param name="row"></param> 
+        public void Append(IEnumerable<object> row = null)
+        {
+            IEnumerator<DataFrameColumn> columnEnumerator = Columns.GetEnumerator();
+            bool columnMoveNext = columnEnumerator.MoveNext();
+            if (row != null)
+            {
+                // Go through row first to make sure there are no data type incompatibilities
+                IEnumerator<object> rowEnumerator = row.GetEnumerator();
+                bool rowMoveNext = rowEnumerator.MoveNext();
+                Dictionary<string, object> cachedObjectConversions = new Dictionary<string, object>();
```

Comment:
You don't need a dictionary here. You just need a `List<object>`. they will be in the right order.

Reply:
Right right, I'm not done yet. I just rebased onto current upstream/master in the middle of doing this